edgelet/support-bundle/src/support_bundle old.rs

```diff
@@ -0,0 +1,684 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use std::env;
+use std::ffi::OsString;
+use std::fs::File;
+use std::io::{Cursor, Seek};
+use std::path::Path;
+use std::process::Command as ShellCommand;
+
+use chrono::{DateTime, Local, NaiveDateTime, Utc};
+use failure::Fail;
+use futures::{Future, Stream};
+use tokio::prelude::*;
+use zip::{write::FileOptions, CompressionMethod, ZipWriter};
+
+use edgelet_core::{LogOptions, LogTail, Module, ModuleRuntime};
+
+use crate::error::{Error, ErrorKind};
+use crate::logs::pull_logs;
+
+pub fn make_bundle<M>(
+    output_location: OutputLocation,
+    log_options: LogOptions,
+    include_ms_only: bool,
+    verbose: bool,
+    iothub_hostname: Option<String>,
+    runtime: M,
+) -> Box<dyn Future<Item = (Box<dyn Read + Send>, u64), Error = Error> + Send>
+where
+    M: 'static + ModuleRuntime + Clone + Send + Sync,
+{
+    match output_location {
+        OutputLocation::File(location) => {
+            let writer = future::result(
+                File::create(Path::new(&location))
+                    .map_err(|err| Error::from(err.context(ErrorKind::SupportBundle))),
+            );
+
+            let state = writer.and_then(move |writer| {
+                make_state(
+                    log_options,
+                    include_ms_only,
+                    verbose,
+                    iothub_hostname,
+                    runtime,
+                    writer,
+                )
+            });
+
+            let bundle = state.and_then(BundleState::bundle_all);
+
+            let read =
+                bundle.and_then(|mut bundle| -> Result<(Box<dyn Read + Send>, u64), Error> {
+                    let result: Box<dyn Read + Send> = Box::new(
+                        bundle
+                            .zip_writer
+                            .finish()
+                            .map_err(|err| Error::from(err.context(ErrorKind::SupportBundle)))?,
+                    );
+                    Ok((result, 0)) // TODO: Get Size
+                });
+
+            Box::new(read)
+        }
+        OutputLocation::Memory => {
+            let writer = future::ok(Cursor::new(Vec::new()));
+
+            let state = writer.and_then(move |writer| {
+                make_state(
+                    log_options,
+                    include_ms_only,
+                    verbose,
+                    iothub_hostname,
+                    runtime,
+                    writer,
+                )
+            });
+
+            let bundle = state.and_then(BundleState::bundle_all);
```

Comment:
Feel free to ignore here too:

I am not sure how much work these make_state calls do... but if it is not much maybe it is okay to not wrap these all as futures and just let them execute normally until we actually have async work to do? If we are using a multi-threaded runtime then I suspect `future::ok(Cursor...)` is an allocation to create a task and requires more cpu cycles to then execute than what would likely normally be elision of Vec::new into the new call.

Reply:
also old file